Subscribe user to newsletter on save

Gregoor opened the pull request:
Fixes #5792 

## What it does
It override the `User`-model's save method to check whether the `salesforce_connection` "enum" is set to `pending` and whether it was empty or `error` before. If those are true it subscribes that user to our newsletter on Salesforce, through [basket](https://basket.readthedocs.io/newsletter_api.html#news-subscribe).

## How to test it

To test this PR, either run the user edit/signup request with curl/Postman/... and add `user-salesforce_connection: 'pending'` to the params or temporarily apply this patch:

<details>
  <summary>user_edit.html patch</summary>

```Index: kuma/users/jinja2/users/user_edit.html
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- kuma/users/jinja2/users/user_edit.html	(revision aa879155482cd67e8b6aee566b8aa7b39dbd9010)
+++ kuma/users/jinja2/users/user_edit.html	(date 1574787500681)
@@ -121,6 +121,13 @@
             {% endif %}
           </li>
 
+            <li class="field">
+                <label for="{{ user_form.salesforce_connection.id_for_label }}">
+                    {{ user_form.salesforce_connection }}
+                    {{ user_form.salesforce_connection.label }}
+                </label>
+            </li>
+
           {{ li_field(user_form, 'twitter_url', classes='site twitter') }}
 
           {{ li_field(user_form, 'stackoverflow_url', classes='site stackoverflow') }}
```
</details>

Now after setting the flag, you should receive an email, welcoming you to the MDN newsletter.

escattone commented on kuma/users/models.py:
```diff
@@ -230,3 +234,9 @@ def get_recovery_url(self):
+    def save(self, *args, **kwargs):
+        old = type(self).objects.get(pk=self.pk) if self.pk else None
+        super().save(*args, **kwargs)
+        if self.is_subscribed and (old is None or not old.is_subscribed):
```
So it looks like we only manage subscriptions here, not un-subscriptions? If we're assuming they can only be done via the unsubscribe link in the newsletter emails, it seems we have no way to keep the `is_subscribed` setting on the user's account synchronized with basket.

Gregoor commented:
Thank you so much for these excellent suggestions, Ryan! I've changed it accordingly. Forgot about our task queue, great to have that. And also the fetching of the previous value is much slimmer & more readable now. yay :)

Also renamed `is_subscribed` to `is_in_salesforce`, I think you were rightfully confused by the naming. I hope this makes more sense now.

I also wondered whether `is_in_salesforce` should be an enum like`SalesforceStatus (None, 'requested', 'subscribed', 'failed')` instead, given that the request might fail (though it looks like django doesn't support mysql enums yet). Then the form would only set the field to `requested` and the task would either set it to `subscribed` or `failed`. 🤔

escattone commented on kuma/users/basket.py:
```diff
@@ -0,0 +1,18 @@
+
+
+@task
+def subscribe(user):
```
I would recommend:
```suggestion
def subscribe(email, username, locale):
```
instead, in order to avoid the serialization/deserialization (i.e., pickling/unpickling) of the user instance when inserted-into/removed-from the Celery task queue, and since the three pieces of information needed are all strings.

Gregoor replied:
Makes sense! No need to waste cycles on uneeded data, will fix. Thanks!

escattone commented:
> I also wondered whether `is_in_salesforce` should be an enum like`SalesforceStatus (None, 'requested', 'subscribed', 'failed')` instead, given that the request might fail (though it looks like django doesn't support mysql enums yet). Then the form would only set the field to `requested` and the task would either set it to `subscribed` or `failed`. 🤔

I like this idea! I suppose we could use the [`choices` option](https://docs.djangoproject.com/en/1.11/ref/models/fields/#choices) on the Field, so something like `choices=[('None', 'None'), ('requested', 'requested'), ('subscribed', 'subscribed'), ('failed', 'failed'))`? Also, I suppose once this value is set to `'subscribed'`, it should never be changed after that?

Gregoor commented:
> I like this idea! I suppose we could use the [`choices` option](https://docs.djangoproject.com/en/1.11/ref/models/fields/#choices) on the Field, so something like `choices=[('None', 'None'), ('requested', 'requested'), ('subscribed', 'subscribed'), ('failed', 'failed'))`?

I was wondering about that; how do we feel about adding a new column that will just have the default value for 99% of the rows? Am I over-optimizing (for disk space, which I suppose is cheap) when I consider moving it into a separate table? Something like:
```mysql
CREATE TABLE salesforce_requests (
  id PRIMARY KEY,
  auth_user_id INT NOT NULL REFERENCES auth_users(id) UNIQUE,
  status ENUM('pending', 'success', 'error') NOT NULL
)

```
(I think this is neither valid postgresql nor valid mysql)

> Also, I suppose once this value is set to `'subscribed'`, it should never be changed after that?

Oh yes, I created the confusion again by calling it `subscribed`. We wouldn't manage that value indeed, hence it also shouldn't be called that 🙊

peterbe commented:
> how do we feel about adding a new column

Some day we might be able to upgrade from mysql to postgres and the less mysql-specifics we use the better. The Django ORM isn't perfect but it's pretty good and if you can stay with what the ORM provides and not have to "go low" into raw SQL I think you're building more sustainably. 

By the way, I much prefer putting Django'esque enums into the model which can be picked up by the Django `forms.ModelForm` classes.

Gregoor commented:
Oh yeah sorry that was misleading, I didn't mean to express that I wanted to use raw SQL to add to the schema, but rather only use it to express a related model with the given fields. But sounds like we're fine with enums, then I'll go with that! Thanks

Gregoor commented:
I've changed it to be an "enum", but I had do to something ugly in the process:
the basket task now has a local import to `User` to work-around circular dependencies not compiling in Python (which is a good thing). But Django turning models into "god" objects is imo not a good thing and makes this hard to solve. I could use signals, but those are imo about as confusing in the long run. Maybe I should just move the save override into a separate file.. 🤔

Gregoor commented:
> 
> 
> I've changed it to be an "enum", but I had do to something ugly in the process:
> the basket task now has a local import to `User` to work-around circular dependencies not compiling in Python (which is a good thing). But Django turning models into "god" objects is imo not a good thing and makes this hard to solve. I could use signals, but those are imo about as confusing in the long run.

@escattone you probably have a full inbox, but if you have a (rough) idea on how to resolve this, that'd be amazing.

peterbe commented on kuma/users/basket.py:
```diff
@@ -0,0 +1,24 @@
+        logger.info(f'Successfully subscribed user {email} to newsletter')
+    except Exception as e:
+        User.objects.filter(pk=pk).update(salesforce_connection='error')
+        raise e
```
```suggestion
        raise
```

peterbe commented on kuma/users/basket.py:
```diff
@@ -0,0 +1,24 @@
+            'lang': locale,
+            'first_name': username
+        }
+        requests.post(BASKET_SUBSCRIPTION_URL, data=data)
```
Also, put a code comment here about something like "# TODO: Use request_retry_session() once https://github.com/mdn/kuma/pull/5424 lands"

peterbe commented:
Perhaps this is not the kind of feedback you'd want to hear, this late in the game, but I'd rather move this to its own model instead of adding yet another string field to the already busy `User` model. E.g.

```python
class User(AbstractUser):
    ... as before, no changes ...


class UserSalesforceConnection(models.model):
    user = models.ForeignKey(User, on_delete=models.CASCADE)
    success = models.BooleanField(default=False)
    pending = models.BooleanField(default=False)
    error = models.TextField(null=True)
    modified = models.DateTimeField(auto_now=True)
    created = models.DateTimeField(auto_add_now=True)
    
```

then you'd get so much more. You'd be able to capture any errors (but you'd have to turn it into a string!) and you'd get a record of when it was attempted, when it was re-attempted, etc. 

You might need to consider not using `models.ForeignKey(User, on_delete=models.CASCADE)` but instead `models.OneToOneField(User, on_delete=models.CASCADE)` if there's only ever 1 of these connections per user ever. 

If you use a `OneToOneFIeld` you'll be able to do smart joins automatically like:
```python
for user in User.objects.all()[:100]:
     try: 
          print(user.usersalesforceconnection.created)
     except UserSalesforceConnection.DoesNotExist:  
         print('No Salesforce connection yet')
```

Either way, if you make it a 0-N connection you'd be able to do something like this:
```python
if not UserSalesforceConnection.objects.filter(user=user, Q(success=True, pending=True)).exists():
    # either we've never tried or we have tried but failed. Just try again!
    ...  
```

(PS. I rush'edly typed the example code, but I think it proves the point)

Gregoor commented on kuma/users/basket.py:
```diff
@@ -0,0 +1,24 @@
+
+@task
+def subscribe(pk, email, username, locale):
+    from .models import User
```
Yes because of circular dependencies, I was hoping to find another solution. But since I didn't I'll add a comment as suggested :+1:

Gregoor commented on kuma/users/forms.py:
```diff
@@ -123,14 +123,18 @@ class UserEditForm(forms.ModelForm):
                   'twitter_url', 'github_url', 'is_github_url_public',
                   'stackoverflow_url', 'linkedin_url', 'mozillians_url',
-                  'facebook_url', 'discourse_url', 'username')
+                  'facebook_url', 'discourse_url', 'username', 'salesforce_connection')
```
Uh leftover from before when the field was a simple user-controlled boolean. That indeed shouldn't be in there!

Gregoor commented:
Thanks for the detailed feedback! Started to put all of that in.

> Perhaps this is not the kind of feedback you'd want to hear, this late in the game, but I'd rather move this to its own model instead of adding yet another string field to the already busy `User` model

I do think it's the better approach, which is why [I was wondering whether I should go with that last week](https://github.com/mdn/kuma/pull/6141#issuecomment-558173260). Given that you've already written half the code in commens, I should probably just go with it 😁

Gregoor commented:
Actually, I do think an enum is good enough, imo it's enough to have the exact error in our logs. Btw, can we set up alerts for specific filter in our logging system?

Wdyt @peterbe, do I get the go ahead?

tobinmori commented:
This is currently blocked by our lack of a SF instance:
https://bugzilla.mozilla.org/show_bug.cgi?id=1562678

Workaround: 
Create a flag in the DB that indicates that a user has opted in to email subscription.

Gregoor commented:
Continued in #6330
